wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsnext/WorkspaceNextObjectsRetriever.java

```diff
@@ -0,0 +1,299 @@
+    if (isNullOrEmpty(featuresAttribute)) {
+      return emptyList();
+    }
+    String[] features = featuresAttribute.split(" *, *");
+    if (features.length == 0) {
+      return emptyList();
+    }
+
+    Collection<Pair<String, String>> featuresNameVersion = parseFeatures(features);
+
+    return toServices(getFeatures(featuresNameVersion));
+  }
+
+  private Collection<Pair<String, String>> parseFeatures(String[] features)
+      throws InfrastructureException {
+    Map<String, Pair<String, String>> featuresNameVersion = new HashMap<>();
+    for (String feature : features) {
+      String[] featureAndVersion = feature.split("/", 2);
+      if (featureAndVersion.length != 2 || featureAndVersion[0].isEmpty() || featureAndVersion[1]
+          .isEmpty()) {
+        throw new InfrastructureException(
+            "Features format is illegal. Problematic feature entry:" + feature);
+      }
+      String key = featureAndVersion[0] + '/' + featureAndVersion[1];
+      if (featuresNameVersion.containsKey(key)) {
+        throw new InfrastructureException(
+            format("Invalid Workspace.Next configuration: feature %s is duplicated", key));
+      }
+      featuresNameVersion.put(key, Pair.of(featureAndVersion[0], featureAndVersion[1]));
+    }
+    return featuresNameVersion.values();
+  }
+
+  private Collection<CheFeature> getFeatures(Collection<Pair<String, String>> featureAndVersion)
+      throws InfrastructureException {
+    List<CheFeature> features = new ArrayList<>();
+    for (Pair<String, String> nameVersion : featureAndVersion) {
+      features.add(getFeature(nameVersion.first, nameVersion.second));
+    }
+
+    return features;
+  }
+
+  private CheFeature getFeature(String featureName, String featureVersion)
+      throws InfrastructureException {
+    try {
+      URI getFeatureURI = featureApi.clone().path("feature").path(featureName).path(featureVersion)
+          .build();
+
+      return getBody(getFeatureURI, CheFeature.class);
+    } catch (IllegalArgumentException | UriBuilderException e) {
+      LOG.error(e.getLocalizedMessage(), e);
+      throw new InternalInfrastructureException(
+          format("Feature %s/%s retrieval failed", featureName, featureVersion));
+    }
+  }
+
+  private Collection<CheService> toServices(Collection<CheFeature> features)
+      throws InfrastructureException {
+    Collection<CheService> services = getServices(features);
+    Map<String, List<String>> parameters = getServicesParameters(features);
+
+    for (CheService service : services) {
+      String serviceName = service.getMetadata().getName();
+      String serviceVersion = service.getSpec().getVersion();
+      String serviceKey = serviceName + '/' + serviceVersion;
+
+      // for now we match whole env variable value against '${<parameter name>}'
+      service.getSpec()
+          .getContainers()
+          .forEach(container -> container.getEnv()
```

Comment:
Maybe it would be better(maybe not =)) to rewrite it in the following manner
```
      service.getSpec()
          .getContainers()
          .stream()
          .flatMap(c -> c.getEnv().stream())
          .forEach(envVar -> {...});
```